## 格式化輸出

## python 3.6 以後新的 f-strings 寫法 (Literal String Interpolation)

f-strings 的寫法為在字串中，插入一組代位符號 {} ，在代位符號內放入變數名稱或運算式。
代位符號可用格式如下：

## {var}  
## {var+var}  
## {var:format}  


In [1]:
import math 

name = "Tom"
age = 3
print(f"His name is {name}, he's {age} years old.")

'''
fstring 運算式
'''
x = 3
y = 5
print(f'x+y={x+y}')

'''
數字格式化 整數保留 5 位，前面補零
'''
se = 2
ss = f"The serial is {se:05d}"
print(ss)

'''
數字格式化 :.2f 精確到小數點以下 2 位
'''
PI = math.pi 
print(f"Pi is {PI:.2f}")

'''
支援生成式
'''
print(f'Numbers from 1-10 are {[_ for _ in range(1, 11)]}')

'''
呼叫自訂函數
'''
def upper(s):
     return s.upper()

stock = 'tsmc'
close = 217.5
print(f'{stock} price: {close}')

print(f'{upper(stock)} price: {close}')

His name is Tom, he's 3 years old.
x+y=8
The serial is 00002
Pi is 3.14
Numbers from 1-10 are [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
tsmc price: 217.5
TSMC price: 217.5


## 進位制的轉換、二進位、16進位、8進位


In [3]:
'''
進位制轉換
'''
x = 31
width = 8  # 數字寬度

print('f-string 寫法: ', '-'*50)
print(f'二進位：{x:0{width}b}')
print(f'8進位：{x:o}')
print(f'16進位：{x:x}')


f-string 寫法:  --------------------------------------------------
二進位：00011111
8進位：37
16進位：1f


## 常見的格式化參數

<table class="code">
<tbody><tr>
<th> Number </th>
<th> Format </th>
<th> Output </th>
<th> Description </th>
</tr>
<tr>
<td> 3.1415926  </td>
<td> {:.2f}  </td>
<td> 3.14  </td>
<td> 2 decimal places </td>
</tr>
<tr>
<td> 3.1415926  </td>
<td> {:+.2f} </td>
<td> +3.14 </td>
<td> 2 decimal places with sign </td>
</tr>
<tr>
<td> -1         </td>
<td> {:+.2f} </td>
<td> -1.00 </td>
<td> 2 decimal places with sign </td>
</tr>
<tr>
<td> 2.71828    </td>
<td> {:.0f}  </td>
<td> 3   </td>
<td> No decimal places </td>
</tr>
<tr>
<td> 5          </td>
<td> {:0&gt;2d} </td>
<td> 05  </td>
<td> Pad number with zeros (left padding, width 2) </td>
</tr>
<tr>
<td> 5          </td>
<td> {:x&lt;4d} </td>
<td> 5xxx  </td>
<td> Pad number with x’s (right padding, width 4) </td>
</tr>
<tr>
<td> 10         </td>
<td> {:x&lt;4d} </td>
<td> 10xx  </td>
<td> Pad number with x’s (right padding, width 4) </td>
</tr>
<tr>
<td> 1000000    </td>
<td> {:,}    </td>
<td> 1,000,000  </td>
<td> Number format with comma separator </td>
</tr>
<tr>
<td> 0.25       </td>
<td> {:.2%}  </td>
<td> 25.00%  </td>
<td> Format percentage </td>
</tr>
<tr>
<td> 1000000000 </td>
<td> {:.2e}  </td>
<td> 1.00e+09 </td>
<td> Exponent notation </td>
</tr>
<tr>
<td> 13         </td>
<td> {:10d}  </td>
<td> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;13 </td>
<td> Right aligned (default, width 10) </td>
</tr>
<tr>
<td> 13         </td>
<td> {:&lt;10d} </td>
<td> 13 </td>
<td> Left aligned (width 10) </td>
</tr>
<tr>
<td> 13         </td>
<td> {:^10d} </td>
<td> &nbsp;&nbsp;&nbsp;&nbsp;13 </td>
<td> Center aligned (width 10) </td>
</tr>
</tbody></table>


①.當被修約的值爲5時，如果他前面的數爲偶數且被修約數的後面沒有數時則捨棄；

②.當被修約的值爲5時，如果他前面的數爲奇數時則進位；

③.當被修約的值爲5時，如果他前面的數爲偶數時且被修約數的後面還有數時，則進位。

有點繁瑣，總之就一個概念“四捨六入五成雙”


In [4]:

x = 29
print(f'固定寬度: {x:7d}')
print(f'固定寬度不足補 0 : {x:05d}')
print(f'固定寬度不足補 0 : {x:05d}')
y = 3.14159
print(f'指定小數點位數為2 : {y:.2f}')

x = 4.5
print(f'指定小數點位數為0 : {x:.0f}')
print(f'四捨六入雙進位 : {1.5:.0f}, {2.5:.0f}, {3.5:.0f}, {2.50001:.0f}')
print(f'顯示為百分比 : {0.45:.0%}')

# 舊式 代位字元
print('The value of PI is approximately %5.3f.' % 3.14159)
print('The value of PI is approximately {:5.3f}.'.format(3.14159))


固定寬度:      29
固定寬度不足補 0 : 00029
固定寬度不足補 0 : 00029
指定小數點位數為2 : 3.14
指定小數點位數為0 : 4
四捨六入雙進位 : 2, 2, 4, 3
顯示為百分比 : 45%
The value of PI is approximately 3.142.
The value of PI is approximately 3.142.


### 亂數

In [4]:
import random

# 產生一個 0 <= randint <=9 之間的一組整數亂數
print(random.randint(0, 9))

# 產生一個 0 <= rand <=1 之間的一組浮點數亂數
print(random.random())

ll = []
for _ in range(100):
    ll.append(random.randint(0, 9999))
    
print(ll)

5
0.06599264777447844
[9985, 3684, 499, 614, 3531, 5395, 2239, 6535, 1039, 2505, 7370, 6572, 8498, 4651, 6027, 6880, 2800, 6521, 6491, 6925, 3482, 2755, 632, 9418, 228, 6553, 5902, 7221, 6826, 863, 6038, 1913, 1686, 4216, 8869, 6105, 8194, 9874, 8398, 6031, 9353, 6887, 7411, 7248, 55, 7243, 1480, 9548, 9040, 1362, 6887, 7657, 2262, 6676, 7783, 1884, 7936, 6793, 8553, 8156, 3918, 756, 7379, 5609, 5306, 9806, 5404, 1005, 2963, 137, 1508, 8552, 3323, 2191, 6558, 5430, 2236, 3965, 8530, 1105, 2937, 6091, 5231, 517, 3474, 8403, 3720, 5320, 4396, 7245, 8111, 1772, 4905, 1941, 1659, 4632, 6381, 4569, 6184, 6358]
